Update Two-phase Commit Transactions document (#644)
scalar-labs/scalardb: docs/two-phase-commit-transactions.md

Currently, you need to call `validate()` when you use the `Consensus Commit` transaction manager with `EXTRA_READ` serializable strategy in `SERIALIZABLE` isolation level.
In other cases, `validate()` does nothing.

> ### Load balancing in Two-phase Commit Transactions

Reviewer: ```suggestion
### Request Routing in Two-phase Commit Transactions
```

> Usually, services using Two-phase Commit Transactions achieve a transaction while exchanging multiple requests/responses as follows:

Reviewer: ```suggestion
Services using Two-phase Commit Transactions usually execute a transaction by exchanging multiple requests and responses as follows:
```

<p align="center">
<img src="images/two_phase_commit_sequence_diagram.png" width="400" />
</p>

> Also, a service typically has multiple hosts/servers for scalability and availability, and uses server-side (proxy) or client-side load balancing to distribute the requests to the hosts/servers.

Reviewer: ```suggestion
Also, each service typically has multiple servers (or hosts) for scalability and availability and uses server-side (proxy) or client-side load balancing to distribute requests to the servers.
```

> Since a transaction processing in Two-phase Commit Transactions is stateful, operations in a transaction must be executed in the same hosts/servers.

Reviewer: ```suggestion
In such a case, since a transaction processing in Two-phase Commit Transactions is stateful, requests in a transaction must be routed to the same servers while different transactions need to be distributed to balance the load.
```

> In other words, the load balancing must distribute the requests in a transaction to the same host/server.

Reviewer: ```suggestion
```

<p align="center">

> <img src="images/two_phase_commit_load_balancing.png" width="500" />

Reviewer 2: In this schema, there is a missing `u` in the `reqest` word placed on the arrow from Client to Service A.

Author: Thank you for pointing it out! I fixed it in 08ba4508825c01587e16fed3b8fc856f368c9ded.

</p>

> As solutions to this issue depend on the protocol between the services, we introduce some solutions for gRPC and HTTP/1.1 here.

Reviewer: ```suggestion
There are several approaches to achieve it depending on the protocol between the services. Here, we introduce some approaches for gRPC and HTTP/1.1.
```

Requests in the same gRPC connection always go to the same host/server in L3/L4 load balancing.
When using an L7 load balancer, since requests in the same gRPC connection do not necessarily go to the same host/server, you need to use cookies or similar for routing requests to correct host/server.
For example, when you use [Envoy](https://www.envoyproxy.io/), you can use session affinity (sticky session) for gRPC.
Or you can also use [Bidirectional streaming RPC in gRPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) since the L7 load balancer distributes requests in the same stream to the same host/server.

#### HTTP/1.1

Typically, you use a server-side (proxy) load balancer with HTTP/1.1.
When using an L3/L4 load balancer, you can use the same HTTP connection to send requests in a transaction, which guarantees the requests go to the same host/server.
When using an L7 load balancer, since requests in the same HTTP connection do not necessarily go to the same host/server, you need to use cookies or similar for routing requests to correct host/server.
You can use session affinity (sticky session) in that case.

> #### Suspend and resume a transaction

Reviewer 3: This is a question. I was just curious.
I could not read a clear reason why I need to suspend the Tx.
Regarding the suspend/resume operations, if I don't suspend Tx as follows, what problems will occur?

```java
public class ServiceBImpl implements ServiceB {

  private TwoPhaseCommitTransactionManager manager = ...;

  ...

  @Override
  public void endpoint1(String txId) throws Exception {
    // First, you need to join the transaction
    TwoPhaseCommitTransaction tx = manager.join(txId);
    ...
  }

  @Override
  public void endpoint2(String txId) throws Exception {
    // You can resume the transaction suspended in endpoint1()
    TwoPhaseCommitTransaction tx = manager.resume(txId);
    ...
  }
...
```

In the above example, will `manager.resume(txId)` in the `endpoint2` return any error? (Since the Tx was not suspended in the previous API `endpoint1`.)

Author: Thank you for taking a look at this!

> In the above example, will manager.resume(txId) in the endpoint2 return any error? (Since the Tx was not suspended in the previous API endpoint1.)

Yes, `manager.resume(txId)` in the `endpoint2` throws an exception in that case.

Reviewer 3: Thank you for your answer!
I understand.
